src/entry.c

```diff
@@ -0,0 +1,363 @@
+    if (server.primary_host == NULL && server.import_mode) {
+        if (server.current_client && server.current_client->flag.import_source) return 0;
+    }
+    return 1;
+}
+/**************************************** Entry Expiry API - End *****************************************/
+
+void entryFree(entry *entry) {
+    if (entryHasValuePtr(entry)) {
+        sdsfree(*entryGetValueRef(entry));
+    }
+    zfree(entryAllocPtr(entry));
+}
+
+/* Takes ownership of value. does not take ownership of field */
+entry *entryCreate(sds field, sds value, long long expiry) {
+    sds embedded_field_sds;
+    size_t expiry_size = (expiry == EXPIRY_NONE) ? 0 : sizeof(long long);
+    size_t field_len = sdslen(field);
+    int field_sds_type = sdsReqType(field_len);
+    if (field_sds_type == SDS_TYPE_5 && (expiry_size > 0)) {
+        field_sds_type = SDS_TYPE_8;
+    }
+    size_t field_size = sdsReqSize(field_len, field_sds_type);
+    size_t value_len = value ? sdslen(value) : 0;
+    size_t embedded_value_size = value ? sdsReqSize(value_len, SDS_TYPE_8) : 0;
+    size_t alloc_size = field_size + expiry_size;
+    bool embed_value = false;
+    if (value) {
+        if (alloc_size + embedded_value_size <= EMBED_VALUE_MAX_ALLOC_SIZE) {
+            /* Embed field and value. Value is fixed to SDS_TYPE_8. Unused
+             * allocation space is recorded in the embedded value's SDS header.
+             *
+             *     +------+--------------+---------------+
+             *     | TTL  | field        | value         |
+             *     |      | hdr "foo" \0 | hdr8 "bar" \0 |
+             *     +------+--------------+---------------+
+             */
+            embed_value = true;
+            alloc_size += embedded_value_size;
+        } else {
+            /* Embed field, but not value. Field must be >= SDS_TYPE_8 to encode to
+             * indicate this type of entry.
+             *
+             *     +------+-------+---------------+
+             *     | TTL  | value | field         |
+             *     |      | ptr   | hdr8 "foo" \0 |
+             *     +------+-------+---------------+
+             */
+            embed_value = false;
+            alloc_size += sizeof(sds *);
+            if (field_sds_type == SDS_TYPE_5) {
+                field_sds_type = SDS_TYPE_8;
+                alloc_size -= field_size;
+                field_size = sdsReqSize(field_len, field_sds_type);
+                alloc_size += field_size;
+            }
+        }
+    }
+    /* allocate the buffer */
+    size_t buf_size;
+    char *buf = zmalloc_usable(alloc_size, &buf_size);
+
+    /* Set The expiry if exists */
+    if (expiry_size) {
+        memcpy(buf, &expiry, expiry_size);
+        buf += expiry_size;
+        buf_size -= expiry_size;
+    }
+    if (value) {
+        if (!embed_value) {
+            *(sds *)buf = value;
+            buf += sizeof(sds *);
+            buf_size -= sizeof(sds *);
```

Comment:
Here, again, I think this is odd because SDS is already a pointer.
Is this more correct?
```suggestion
            (sds)buf = value;
            buf += sizeof(sds);
            buf_size -= sizeof(sds);
```

Reply:
agree. but for the:
```
(sds)buf = value;
```

this will not do. I need to change the pointer value stored inside the buffer...